actions.py

```diff
@@ -0,0 +1,62 @@
+import pickle
+import base64
+from pathlib import Path
+import subprocess
+
+from flask import Blueprint, request, jsonify, session
+
+bp = Blueprint("actions", __name__)
+
+
+@bp.route("/message", methods=["POST"])
+def log_entry():
+    user_info = session.get("user_info", None)
+    if user_info is None:
+        return jsonify({"error": "no user_info found in session"})
+    access_level = user_info[2]
+    if access_level > 2:
+        return jsonify({"error": "access level < 2 is required for this action"})
+    filename_param = request.form.get("filename")
+    if filename_param is None:
+        return jsonify({"error": "filename parameter is required"})
+    text_param = request.form.get("text")
+    if text_param is None:
+        return jsonify({"error": "text parameter is required"})
+
+    user_id = user_info[0]
+    user_dir = "data/" + str(user_id)
+    user_dir_path = Path(user_dir)
+    if not user_dir_path.exists():
+        user_dir_path.mkdir()
+
+    filename = filename_param + ".txt"
+    path = Path(user_dir + "/" + filename)
```

Comment:
<div align="center"> 

### 🎯 Static Application Security Testing (SAST) Vulnerability 
	
| Severity | Finding |
| :---: | :---: |
| ![](https://raw.githubusercontent.com/jfrog/frogbot/master/resources/v2/applicableHighSeverity.png)<br> | Path Traversal |

</div>

<details>
<summary> <b>Full description</b> </summary>
<br>


### Overview
Path traversal, also known as directory traversal, is a type of vulnerability that allows an
attacker to access files or directories on a computer or device that are outside of
the intended directory.
Allowing arbitrary read access can allow the attacker to read sensitive files, such as
configuration files or sensitive data, potentially leading data loss
or even system compromise.
Allowing arbitrary write access is more severe and in most cases leads to arbitrary code
execution, via editing important system files or sensitive data.

### Query operation
In this query we look for user input that can flow un-sanitized as a path into file access
functions
(either read or write access)

### Vulnerable example
```python
from flask import Flask, request, send_file
app = Flask(__name__)

@app.route('/files/')
def serve_file():
    filename = request.args.get('filename')
    basepath = 'static/files/'
    return send_file(basepath + filename)

if __name__ == '__main__':
    app.run()
```
In this example, the application has a route `/files/` that serves files from a directory
called `static/files`. The vulnerability lies in the fact that the application does not
properly validate or sanitize the `filename` parameter, allowing an attacker to traverse
beyond the intended directory and access sensitive files on the server.
An attacker could exploit this vulnerability by manipulating the `filename` parameter
and providing a relative path to access files outside of the `static/files` directory.
For example, they could craft a URL like this:
`http://localhost:5000/files/?filename=../../../etc/passwd`

### Remediation
When possible, use inherently safe path functions such as `send_from_directory` that perform
filename escaping -
```diff
@app.route('/files/')
def serve_file():
    filename = request.args.get('filename')
    basepath = 'static/files/'
-    return send_file(basepath + filename)
+    return send_from_directory(basepath, filename)
```
Alternatively, before accessing a potential path, check that the user's `filename` does not
escape the intended path -
```python
from pathlib import Path
def is_escaping_path(basepath, userpath):
    try:
        Path(basepath).joinpath(userpath).resolve().relative_to(basepath.resolve())
        return False
    except ValueError:
        return True
```

```diff
@app.route('/files/')
def serve_file():
    filename = request.args.get('filename')
    basepath = 'static/files/'
+    if is_escaping_path(basepath, filename):
+         abort(400)  # Bad Request
    return send_file(basepath + filename)
```


</details>



<details>
<summary><b>Code Flows</b> </summary>



<details>
<summary><b>Vulnerable data flow analysis result</b> </summary>
<br>

↘️. `request.form` (at actions.py line 19)

↘️. `request.form.get` (at actions.py line 19)

↘️. `request.form.get("filename")` (at actions.py line 19)

↘️. `filename_param` (at actions.py line 19)

↘️. `filename_param + ".txt"` (at actions.py line 32)

↘️. `filename` (at actions.py line 32)

↘️. `user_dir + "/" + filename` (at actions.py line 33)

↘️. `Path(user_dir + "/" + filename)` (at actions.py line 33)


</details>



</details>


---
<div align="center">

[🐸 JFrog Frogbot](https://github.com/jfrog/frogbot#readme)

</div>